ml-proto/spec/float.ml

```diff
@@ -0,0 +1,181 @@
+    if xf = 0.0 then x else
+    (* trunc is either ceil or floor depending on which one is toward zero *)
+    let f = if xf < 0.0 then Pervasives.ceil xf else Pervasives.floor xf in
+    bits_of_arith f
+
+  let nearest x =
+    let xf = arith_of_bits x in
+    (* preserve the sign of zero *)
+    if xf = 0.0 then x else
+      (* nearest is either ceil or floor depending on which is nearest or even *)
+    let u = Pervasives.ceil xf in
+    let d = Pervasives.floor xf in
+    let um = abs_float (xf -. u) in
+    let dm = abs_float (xf -. d) in
+    let u_or_d = um < dm ||
+                 (um = dm && let h = u /. 2. in Pervasives.floor h = h) in
+    let f = if u_or_d then u else d in
+    bits_of_arith f
+
+  let min x y =
+    let xf = arith_of_bits x in
+    let yf = arith_of_bits y in
+    (* min -0 0 is -0 *)
+    if xf = yf then Rep.logor x y else
+    if xf < yf then x else
+	if xf > yf then y else
+	nondeterministic_nan
+
+  let max x y =
+    let xf = arith_of_bits x in
+    let yf = arith_of_bits y in
+    (* max -0 0 is 0 *)
+    if xf = yf then Rep.logand x y else
+    if xf > yf then x else
+	if xf < yf then y else
+	nondeterministic_nan
+
+  (* abs, neg, and copysign are purely bitwise operations, even on NaN values *)
+  let abs x =
+    Rep.logand x Rep.max_int
+
+  let neg x =
+    Rep.logxor x Rep.min_int
+
+  let copysign x y =
+    Rep.logor (abs x) (Rep.logand y Rep.min_int)
+
+  let eq x y = (arith_of_bits x) =  (arith_of_bits y)
+  let ne x y = (arith_of_bits x) <> (arith_of_bits y)
+  let lt x y = (arith_of_bits x) <  (arith_of_bits y)
+  let gt x y = (arith_of_bits x) >  (arith_of_bits y)
+  let le x y = (arith_of_bits x) <= (arith_of_bits y)
+  let ge x y = (arith_of_bits x) >= (arith_of_bits y)
+
+  let of_signless_string x len =
+    if x <> "nan" &&
+         (len > 7) &&
+           (String.sub x 0 6) = "nan(0x" && (String.get x (len - 1)) = ')' then
+      (let s = Rep.of_string (String.sub x 4 ((len - 5))) in
+       if s = Rep.zero then
+         raise (Failure "nan payload must not be zero")
+       else if Rep.logand s bare_nan <> Rep.zero then
+         raise (Failure "nan payload must not overlap with exponent bits")
+       else if s < Rep.zero then
+         raise (Failure "nan payload must not overlap with sign bit")
+       else
+         Rep.logor s bare_nan)
+    else
+      (* TODO: OCaml's float_of_string is insufficient *)
+      of_float (float_of_string x)
+
+  let of_string x =
+    let len = String.length x in
+    if len > 0 && (String.get x 0) = '-' then
+      neg (of_signless_string (String.sub x 1 (len - 1)) (len - 1))
+    else if len > 0 && (String.get x 0) = '+' then
+      of_signless_string (String.sub x 1 (len - 1)) (len - 1)
+    else
+      of_signless_string x len
+
+  let to_string x =
+    (if x < Rep.zero then "-" else "") ^
+      let a = abs x in
+      let af = arith_of_bits a in
+      if af <> af then
+        ("nan(0x" ^ Rep.print_nan a ^ ")")
```

Comment:
As a minor nit, `print_nan` here should be named something like `print_nan_significand_digits` since it's just printing that one part, rather than the whole NaN.